components/service-operator/controllers/serviceaccount.go

```diff
@@ -0,0 +1,153 @@
+		"after", res.RequeueAfter,
+		"op", op,
+		"err", err,
+	)
+	return res, err
+}
+
+// reconcileWithContext fetches the resource to reconcile and executes reconcileObjectWithContext and returns if any changes were made
+func (r *ServiceAccountController) reconcileWithContext(ctx context.Context, req ctrl.Request) (controllerutil.OperationResult, error) {
+	bg := context.Background()
+	o := &core.ServiceAccount{}
+	if err := r.KubernetesClient.Get(bg, req.NamespacedName, o); apierrs.IsNotFound(err) {
+		// nothing we can do if the resource has gone missing, so
+		// ignore any not found errors and let the api carry on
+		return controllerutil.OperationResultNone, nil
+	} else if err != nil {
+		// issue communicating with the api
+		// return err and we'll retry later
+		return controllerutil.OperationResultNone, err
+	}
+
+	if _, ok := o.ObjectMeta.Labels[cloudformation.AccessGroupLabel]; !ok {
+		return controllerutil.OperationResultNone, nil
+	}
+
+	// track changes to our object resource and call the main reconcile func
+	var reconcileErr error
+	op, updateErr := controllerutil.CreateOrUpdate(bg, r.KubernetesClient, o, func() error {
+		reconcileErr = r.reconcileObjectWithContext(ctx, req, o)
+		return nil // always try to update
+	})
+	if reconcileErr != nil {
+		return op, reconcileErr
+	} else if updateErr != nil {
+		return op, updateErr
+	}
+	return op, nil
+}
+
+func (r *ServiceAccountController) updatePrincipal(ctx context.Context, o *core.ServiceAccount) (*access.Principal, error) {
+	principal := &access.Principal{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      o.GetName(),
+			Namespace: o.GetNamespace(),
+		},
+	}
+	principalKey, err := client.ObjectKeyFromObject(principal)
+	if err != nil {
+		return nil, err
+	}
+	err = r.KubernetesClient.Get(ctx, principalKey, principal)
+	if err != nil && !apierrs.IsNotFound(err) {
+		return nil, err
+	}
+	op, err := controllerutil.CreateOrUpdate(ctx, r.KubernetesClient, principal, func() error {
+		principal.ObjectMeta.Labels = o.ObjectMeta.Labels
+		// mark the principal as owned by the o resource so it gets gc'd
+		if err := controllerutil.SetControllerReference(o, principal, r.Scheme); err != nil {
+			return err
+		}
+		return nil
+	})
+	r.Log.Info("update-principal",
+		"principal", principalKey,
+		"op", op,
+		"err", err,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return principal, nil
+}
+
+// reconcileObjectWithContext is the main loop, it will mutate "o" with any changes required
+func (r *ServiceAccountController) reconcileObjectWithContext(ctx context.Context, req ctrl.Request, o *core.ServiceAccount) error {
+	defer r.Log.Info("reconcileObjectWithContext",
+		"o", o,
+	)
+	// examine DeletionTimestamp to determine if object is under deletion
+	if !o.GetDeletionTimestamp().IsZero() {
+		// The object is being deleted
+		return nil
+	}
+
+	principal, err := r.updatePrincipal(ctx, o)
+	if err != nil {
+		return err
+	}
+
+	if principal.Status.State != "READY" {
```

Comment:
could we use the `object.ReadyState` constant here instead of the string to let the type system help us avoid typos